service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/credential/cache/CachedPasswordMatcher.java

```diff
@@ -0,0 +1,53 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eurotech and/or its affiliates and others
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Eurotech - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.kapua.service.authentication.credential.cache;
+
+import java.util.Base64;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.eclipse.kapua.commons.cache.Cache;
+import org.eclipse.kapua.commons.cache.LocalCache;
+import org.eclipse.kapua.service.authentication.credential.Credential;
+import org.eclipse.kapua.service.authentication.shiro.setting.KapuaAuthenticationSetting;
+import org.eclipse.kapua.service.authentication.shiro.setting.KapuaAuthenticationSettingKeys;
+import org.springframework.security.crypto.bcrypt.BCrypt;
+
+public class CachedPasswordMatcher implements PasswordMatcher {
```

Comment:
this is a layer introduced just to allow a specific Shiro realm to enable/disable depending on the configuration parameter so there is no need to intriduce other layers since we are using only BCrypt algorithm and this is a specific (shiro) authentication/autorization module.
If you inject here a different algorithm you will break the functionality of this module since the BCrypt is used in other parts of this module.

Reply:
The rest of the code being heavily dependent on a specific algorithm is bad enough, I agree. 
I'd still prefer to see a reusable caching wrapper which need to know nothing of the underlying mechanism - otherwise we are just compounding the problem (the day we need to change the encryption we need to change the invocation to BCrypt both here AND in the Default implementation.